src/puppetlabs/puppetdb/cli/services.clj

```diff
@@ -357,17 +353,23 @@
   that trapperkeeper will call on exit."
   PuppetDBServer
   [[:ConfigService get-config]
-   [:WebroutingService add-ring-handler get-route]]
-
+   [:WebroutingService add-ring-handler get-registered-endpoints]]
+  (init [this context]
+        (assoc context :url-prefix (atom nil)))
   (start [this context]
-         (start-puppetdb context (get-config) this add-ring-handler get-route))
+         (start-puppetdb context (get-config) this get-registered-endpoints))
 
   (stop [this context]
         (stop-puppetdb context))
+  (set-url-prefix [this url-prefix]
+                  (let [old-url-prefix (:url-prefix (service-context this))]
+                    (when-not (compare-and-set! old-url-prefix nil url-prefix)
+                      (throw (RuntimeException. (format "Attempt to set url-prefix to %s when it's already been set to %s" url-prefix @old-url-prefix))))))
```

Comment:
I prefer the slingshot exceptions, but used runtime here because I thought that's what most of our user facing exceptions used. Have we made the switch to slingshot? Happy to update it if that's the case.